README.md
# Vatsim Online

A Ruby gem for selectively pulling, parsing and displaying Vatsim online
stations data. Essentially it's a "Who's online" library, capable of displaying
online ATC and/or pilots for given airports, areas or globally and to filter by callsign. Stations are
returned as objects, exposing a rich set of attributes. Vatsim data is pulled
on preset intervals and cached locally to avoid flooding the servers.

[![Build Status](https://secure.travis-ci.org/tarakanbg/vatsim_online.png?branch=master)](http://travis-ci.org/tarakanbg/vatsim_online)
[![Gemnasium](https://gemnasium.com/tarakanbg/vatsim_online.png?travis)](https://gemnasium.com/tarakanbg/vatsim_online)
[![Gem Version](https://badge.fury.io/rb/vatsim_online.png)](http://badge.fury.io/rb/vatsim_online)
[![Code Climate](https://codeclimate.com/github/tarakanbg/vatsim_online.png)](https://codeclimate.com/github/tarakanbg/vatsim_online)

## Requirements

[Ruby 2.0](http://www.ruby-lang.org/en/downloads/) or higher

## Installation

Add this line to your application's Gemfile:

    gem 'vatsim_online'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install vatsim_online

## Usage

This gem provides two public methods:

* `vatsim_online`, which can be applied to
any string (or variable containing a string) representing a full or partial ICAO
code or a comma-separated list of ICAO codes. The provided ICAO code or fragment
will be used as a search criteria and matched against the current vatsim data.

* `vatsim_callsign`, which can be applied to
any string (or variable containing a string) representing a full or partial pilot or ATC
callsign or a comma-separated list of callsigns. The provided callsign or fragment
will be used as a search criteria and matched against the current vatsim data.

### Filter by ICAO code examples

For example if you want to retrieve all active stations (ATC positions and pilots)
for Vienna airport (ICAO: LOWW), then you can use:

```ruby
# Attaching the method directly to a string:
"LOWW".vatsim_online

# Attaching the method directly to a list of ICAOs:
"LOWW, LOWK".vatsim_online

# Attaching the method to a variable containing a string:
icao = "LOWW"
icao.vatsim_online
```
If you want to retrieve the currently active stations for an entire region
(FIR/ARTCC), then you can use the first 2-3 letters of the region's ICAO name.
For example if you want to pull all the stations active in Austria (ICAO code
for the FIR is LOVV), you can use `"LO"` as your ICAO search string: all Austrian
airports and ATC station callsigns start with `"LO"`:

```ruby
# Attaching the method directly to a string:
"LO".vatsim_online

# Attaching the method to a variable containing a string:
icao = "LO"
icao.vatsim_online

# Attaching the method to a list of ICAOs:
icao = "LO, LB"
icao.vatsim_online
```

When parsing the pilot stations for particular airport or area, the library will
return the pilots that are flying **to or from** the given area or airport,
not the current enroute stations. The discovery algorithm is based on **origin
and destination**.

### Filter by callsign examples

```ruby
# Attaching the method directly to a string:
"BAW".vatsim_callsign # => returns an array of all Speedbird flights as station objects

# Attaching the method to a variable containing a string:
callsign = "BAW"
callsign.vatsim_callsign # => returns an array of all Speedbird flights as station objects

# Attaching the method to a list of callsigns:
callsign = "BAW, RYR"
callsign.vatsim_callsign # => returns an array of all Speedbird and Ryanair flights as station objects
```


### Anatomy of method returns

The `vatsim_online` method returns a **hash** of 4 elements: 1) the matching *atc*
stations, 2) all matching *pilots* stations, 3) matching *arrivals*, 4) matching
*departures*. Each of those is an **array**, cosnsisting of
station **objects**. Each of these objects includes a number of **attributes**:

```ruby
icao.vatsim_online # => {:atc => [a1, a2, a3 ...], :pilots => [p1, p2, p3, p4 ...],
                   #    :departures => [p1, p4 ...], :arrivals => [p2, p3...]}

icao.vatsim_online[:atc] #=> [a1, a2, a3 ...]
icao.vatsim_online[:pilots] #=> [p1, p2, p3 ...]
icao.vatsim_online[:departures] #=> [p1, p3 ...]
icao.vatsim_online[:arrivals] #=> [p2, p4 ...]


icao.vatsim_online[:atc].first #=> a1
icao.vatsim_online[:pilots].first #=> p1

a1.callsign #=> "LQSA_TWR"
a1.frequency #=> "118.25"
a1.name #=> "Svilen Vassilev"
a1.rating #=> "S2"
a1.online_since #=> "2012-09-22 10:00:48 UTC"
...

p1.callsign #=> "ACH217S"
p1.departure #=> "LQSA"
p1.destination #=> "LDSP"
p1.remarks #=> "/V/ RMK/CHARTS"
...
```

`Arrivals` and `departures` are returned separately for convenience, in case you
want to loop through them separately. The `pilots` array contains all arrivals
and departures.

The `vatsim_callsign` method returns an **array** of all matching stations.

### Station attributes

Here's a complete list of the station object attributes that can be accessed:

* `cid` (VATSIM ID)
* `callsign`
* `name`
* `role`
* `frequency`
* `altitude`
* `planned_altitude` (or FL)
* `heading`
* `groundspeed`
* `transponder`
* `aircraft`
* `origin`
* `destination`
* `route`
* `rating` (returns a humanized version of the VATSIM rating: S1, S2, S3, C1, etc...)
* `facility`
* `remarks`
* `atis` (raw atis, as reported from VATSIM, including voice server as first line)
* `atis_message` (a humanized version of the ATC atis w/o server and with lines split)
* `logon` (login time as unparsed text string: `20120722091954`)
* `online_since` (returns the station login time parsed as a Ruby Time object in UTC)
* `latitude`
* `longitude`
* `latitude_humanized` (e.g. N44.09780)
* `longitude_humanized` (e.g. W58.41483)
* `qnh_in` (indicated QNH in inches Hg)
* `qnh_mb` (indicated QNH in milibars/hectopascals)
* `flight_type` (`I` for IFR, `V` for VFR, etc)
* `gcmap` (returns a great circle map image url)

### Great circle map visualization

The `.gcmap` method, available for all pilot stations, returns a great
circle map image url, depicting the GC route of the aircraft, its origin and destination,
its current position on the route, together with its current altitude and groundspeed.
Example below:

```ruby
icao.vatsim_online[:pilots].first.gcmap #=> image url of the map
```

![GC Map](http://www.gcmap.com/map?P=kdfw-N44.09780+W58.41483-egll,+%22AAL026%5cn37112+ft%5cn516+kts%22%2b%40N44.09780+W58.41483&MS=wls&MR=540&MX=720x360&PM=*)

The map size (width and height) and scale **can be customized** by passing an
optional hash of arguments to the `vatsim_online` method like this:

```ruby
icao.vatsim_online(:gcmap_width => "400", :gcmap_height => "400")[:pilots].first.gcmap #=> image url of the map

# The quotes are optional, so the statement can also be written like this:
icao.vatsim_online(:gcmap_width => 400, :gcmap_height => 400)[:pilots].first.gcmap
```
![GC Map](http://www.gcmap.com/map?P=kdfw-N44.09780+W58.41483-egll,+%22AAL026%5cn37112+ft%5cn516+kts%22%2b%40N44.09780+W58.41483&MS=wls&MR=540&MX=400x400&PM=*)

### Customizing the request

The `vatsim online` method can be customized by passing in a hash-style collection
of arguments. The currently supported arguments and their defaults are:

```ruby
:atc => true     # Possible values: true, false. Default: true
:pilots => true  # Possible values: true, false. Default: true
:exclude => "ICAO" # Accepts any full or partial ICAO code to be excluded from the ATC stations
:gcmap_width => integer # Optional parameter customizing the width of the station's `.gcmap`
:gcmap_height => integer # Optional parameter customizing the height of the station's `.gcmap`
```
#### ATC exclusions

The `:exclude => "ICAO"` option can be used to exclude matching stations from the
ATC listing. It accepts full or partial ICAO codes or callsigns.

For example you might want to display all Austrian ATC stations by calling
`"LO".vatsim_online`. Seems logical since all Austrian stations callsigns begin
with the `LO` prefix. However a few UK stations (London Control for example) also
have callsigns beginning with `LO` (*LON_CTR*). To avoid including them, you can
use the `:exclude` option on your request like this:

```ruby
# Lets exclude all ATC station names beginning with LON from our request
"LO".vatsim_online(:exclude => "LON")[:atc] #=> [atc1, atc2, atc3...]

# This option is not case sensitive so we can use "lon" as well:
"LO".vatsim_online(:exclude => "lon")[:atc] #=> [atc1, atc2, atc3...]
```

#### ATC and PILOTS options

Both options can be used to exclude **all** ATC or pilots stations respectively from
the request, in order to speed it up and avoid processing useless data.

**Examples:**

```ruby
# Lets exclude all ATC from our request and get the pilots only
"LO".vatsim_online(:atc => false)[:pilots] #=> [p1, p2, p3...]

# Lets exclude all pilots from our request and get the ATC only
"LO".vatsim_online(:pilots => false)[:atc] #=> [a1, a2, a3...]

"LO".vatsim_online(:atc => false)[:pilots].first.callsign #=> "ACH0838"
"LO".vatsim_online(:pilots => false)[:atc].first.callsign #=> "LOVV_CTR"

```

### Example of Ruby on Rails implementation

Here's a possible scenario of using this gem in a Ruby on Rails application.
Verbosity is kept on purpose for clarity.

**In your controller:**
```ruby
def index
  # We want to retrieve all Austrian online stations (ATC and pilots)
  icao = "LO"
  stations = icao.vatsim_online

  # Now we will assign the ATCs and the pilots to separate instance variables,
  # to be able to loop through them separately in the view
  @atc = stations[:atc]
  @pilots = stations[:pilots]

  # We can also isolate the departures and/or arrivals for conveneinence if we
  # want to loop through them separately
  @departures = stations[:departures]
  @arrivals = stations[:arrivals]
end
```

**In your view (HAML is used for clarity):**

```haml
- for atc in @atc
  %li
    = atc.callsign
    = atc.frequency
    = atc.rating
    = atc.name
    = atc.online_since
    = atc.atis_message

- for pilot in @pilots
  %li
    = pilot.callsign
    = pilot.name
    = pilot.origin
    = pilot.destination
    = pilot.route
    = pilot.flight_type
    = pilot.altitude
    = pilot.groundspeed
    = pilot.heading
    = pilot.remarks
    = pilot.online_since
    = image_tag pilot.gcmap

- for arrival in @arrivals
  %li
    = arrival.callsign
    = arrival.name
    = image_tag arrival.gcmap

- for departure in @departures
  %li
    = departure.callsign
    = departure.name
    = image_tag departure.gcmap
```

### Notes

* Vatsim status and data files are cached locally to reduce the load on vatsim
servers. Random server is chosen to retrieve the data each time. By default the
status file is updated once every 4 hours and the data file once every 3 minutes
regardless of the number of incoming requests.
* The data is cached in your default TEMP directory (OS specific)
* All the data retrieval and caching logic is encapsulated in a separate class
`VatsimTools::DataDownloader` which can be mixed in other applications and
libraries too.
* The ICAO string used as a search criteria **is not** case sensitive
* Pilot stations returned are based on origin and destination airports, the
current algorithm does not evaluate enroute flights.
* When attaching the `vatsim_online` method to a comma-separated list of full or
partial ICAO identifiers it does not matter whether there will be any spaces in
front or after the identifiers or the commas, i.e. you can use
`"LO,LB".vatsim_online` or `"LO, LB".vatsim_online` or `"LO , LB".vatsim_online` with
the same result.

## Changelog

### v. 0.9.1 - 31 Dec 2016

* Maintenance release, updated dependencies

### v. 0.9 - 20 June 2015

* Fixed nil String bug, can select "ALL" (thanks to Pierre Ferran and Florian Rimoli)

### v. 0.8.3 - 10 April 2014

* Fix stale data bug, change data fallback logic

### v. 0.8.2 - 5 October 2013

* Handle undetermined position when drawing gcmap

### v. 0.8.1 - 5 October 2013

* Ignore prefile data

### v. 0.8 - 5 October 2013

* search by callsign implemented

### v. 0.7.4 - 2 August 2013

* fallback scenarios for offline data servers

### v. 0.7.2 - 15 July 2013

* added gemspec license declaration

### v. 0.7.1 - 15 July 2013

* more comprehensive ATC ratings list
* dependencies updated

### v. 0.7.0 - 26 February 2013

* added Ruby 2.0 support

### v. 0.6.2 - 4 January 2013

* updated gem dependencies
* refactored `Station` class for simplicity

### v. 0.6.1 - 08 October 2012

* Fixed pilot station duplication issue when using multiple ICAOs

### v. 0.6.0 - 08 October 2012

* The `vatsim_online` method now also supports a comma-separated list of full or
partial ICAO codes like this: `"LO,LB".vatsim_online`. This allows you to pull the
information for multiple airports or FIRs in a single request
* The comma-seprated list is not sensitive to whitespace, meaning you can use
`"LO,LB".vatsim_online` or `"LO, LB".vatsim_online` or `"LO , LB".vatsim_online` with
the same result

### v. 0.5.3 - 30 September 2012

* fixed bug with exceptions on missing ATC remark

### v. 0.5.2 - 29 September 2012

* fixed permissions bug on UNIX systems

### v. 0.5.1 - 23 September 2012

* bugfixes

### v. 0.5 - 23 September 2012

* New option `:exclude => "ICAO"` allowing further request customization by
excluding a matching subset of ATC stations from the listing. Read the documentation
for detailed explanation and examples
* New customized station attribute: `online_since`. Returns the station login time
parsed as a Ruby Time object in UTC (zulu time). As opposed to the `logon` attribute
which returns an unformatted, unparsed string such as `20120722091954`
* The `rating` station attribute is now humanized not to return just an integer,
but a readable version of the VATSIM rating, i.e. S1, S2, S3, C1, C3, I1, I3, etc...
* Added the possibility of customizing the great circle maps by optionally passing
parameters for width and height: `icao.vatsim_online(:gcmap_width => 400, :gcmap_height => 400)`.
Read the documentation for detailed explanation and examples
* New customized station attribute `atis_message`. It will return a humanized web safe
version of the ATC atis without the voice server info and with lines split with
`<br />` tags. As opposed to the original `atis` attribute, which returns raw atis,
as reported from VATSIM, including voice server as first line

### v. 0.4 - 27 August 2012

* GCMapper integration: this library now plays nicely with [gcmapper](https://rubygems.org/gems/gcmapper).
A new attribute is provided for the pilot stations: `.gcmap` which returns a great
circle map image url, depicting the GC route of the aircraft, its origin and destination,
its current position on the route, together with its current altitude and groundspeed.
Look at the example in the README section above.
* New station attributes: planned_altitude, transponder, heading, qnh_in, qnh_mb,
flight_type, cid, latitude_humanized, longitude_humanized


### v. 0.3 - 22 July 2012

* The hash returned by the `vatsim_online` method now includes 2 new arrays:
`arrivals` and `departures`. These two are returned separately for convenience,
in case you want to loop through them separately. The `pilots` array return is
unchanged and contains all arrivals and departures.
* New station attributes: latitude, longitude
* Improved UTF-8 conversion process

### v. 0.2 - 21 July 2012

* Station attribute `departure` is now renamed to `origin`
* UTF-8 is now enforced for all local caching and file/string manipulations, the
original Vatsim data is re-encoded
* Station ATIS is now cleaned of invalid and obscure characters
* Improved documentation

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Make sure all tests are passing!
5. Push to the branch (`git push origin my-new-feature`)
6. Create new Pull Request

## Credits

Copyright © 2016 [Svilen Vassilev](http://svilen.rubystudio.net)

### Contributors

* [Florian Rimoli](https://github.com/Flox06)
* [Pierre Ferran](https://github.com/pierr3)

Released under the [MIT LICENSE](https://github.com/tarakanbg/vatsim_online/blob/master/LICENSE)

Maps generated by the [Great Circle Mapper](http://www.gcmap.com/), copyright Karl L. Swartz


CHANGELOG.md
# vatsim_online

## Changelog

### v. 0.9.1 - 31 Dec 2016

* Maintenance release, updated dependencies

### v. 0.9 - 20 June 2015

* Fixed nil String bug, can select "ALL" (thanks to Pierre Ferran and Florian Rimoli)

### v. 0.8.3 - 10 April 2014

* Fix stale data bug, change data fallback logic

### v. 0.8.2 - 5 October 2013

* Handle undetermined position when drawing gcmap

### v. 0.8.1 - 5 October 2013

* Ignore prefile data

### v. 0.8 - 5 October 2013

* search by callsign implemented

### v. 0.7.4 - 2 August 2013

* fallback scenarios for offline data servers

### v. 0.7.2 - 15 July 2013

* added gemspec license declaration

### v. 0.7.1 - 15 July 2013

* more comprehensive ATC ratings list
* dependencies updated

### v. 0.7.0 - 26 February 2013

* added Ruby 2.0 support

### v. 0.6.2 - 4 January 2013

* updated gem dependencies
* refactored `Station` class for simplicity

### v. 0.6.1 - 08 October 2012

* Fixed pilot station duplication issue when using multiple ICAOs

### v. 0.6.0 - 08 October 2012

* The `vatsim_online` method now also supports a comma-separated list of full or
partial ICAO codes like this: `"LO,LB".vatsim_online`. This allows you to pull the
information for multiple airports or FIRs in a single request
* The comma-seprated list is not sensitive to whitespace, meaning you can use
`"LO,LB".vatsim_online` or `"LO, LB".vatsim_online` or `"LO , LB".vatsim_online` with
the same result

### v. 0.5.3 - 30 September 2012

* fixed bug with exceptions on missing ATC remark

### v. 0.5.2 - 29 September 2012

* fixed permissions bug on UNIX systems

### v. 0.5.1 - 23 September 2012

* bugfixes

### v. 0.5 - 23 September 2012

* New option `:exclude => "ICAO"` allowing further request customization by
excluding a matching subset of ATC stations from the listing. Read the documentation
for detailed explanation and examples
* New customized station attribute: `online_since`. Returns the station login time
parsed as a Ruby Time object in UTC (zulu time). As opposed to the `logon` attribute
which returns an unformatted, unparsed string such as `20120722091954`
* The `rating` station attribute is now humanized not to return just an integer,
but a readable version of the VATSIM rating, i.e. S1, S2, S3, C1, C3, I1, I3, etc...
* Added the possibility of customizing the great circle maps by optionally passing
parameters for width and height: `icao.vatsim_online(:gcmap_width => 400, :gcmap_height => 400)`.
Read the documentation for detailed explanation and examples
* New customized station attribute `atis_message`. It will return a humanized web safe
version of the ATC atis without the voice server info and with lines split with
`<br />` tags. As opposed to the original `atis` attribute, which returns raw atis,
as reported from VATSIM, including voice server as first line

### v. 0.4 - 27 August 2012

* GCMapper integration: this library now plays nicely with [gcmapper](https://rubygems.org/gems/gcmapper).
A new attribute is provided for the pilot stations: `.gcmap` which returns a great
circle map image url, depicting the GC route of the aircraft, its origin and destination,
its current position on the route, together with its current altitude and groundspeed.
Look at the example in the README section above.
* New station attributes: planned_altitude, transponder, heading, qnh_in, qnh_mb,
flight_type, cid, latitude_humanized, longitude_humanized


### v. 0.3 - 22 July 2012

* The hash returned by the `vatsim_online` method now includes 2 new arrays:
`arrivals` and `departures`. These two are returned separately for convenience,
in case you want to loop through them separately. The `pilots` array return is
unchanged and contains all arrivals and departures.
* New station attributes: latitude, longitude
* Improved UTF-8 conversion process

### v. 0.2 - 21 July 2012

* Station attribute `departure` is now renamed to `origin`
* UTF-8 is now enforced for all local caching and file/string manipulations, the
original Vatsim data is re-encoded
* Station ATIS is now cleaned of invalid and obscure characters
* Improved documentation


spec/spec_helper.rb
require 'vatsim_online'




spec/vatsim_online_spec.rb
require 'spec_helper.rb'

describe VatsimOnline do

  describe "vatsim_online" do
    it "should work :)" do
      gem_data_file
      "ZGGG".vatsim_online.class.should eq(Hash)
      "LO".vatsim_online.class.should eq(Hash)
      "WM".vatsim_online[:atc].size.should eq(4)
      "LO".vatsim_online[:pilots].size.should eq(12)
      "LO".vatsim_online(:pilots => true, :atc => true).class.should eq(Hash)
      "LO".vatsim_online(:pilots => true, :atc => true)[:atc].size.should eq(0)
      "LO".vatsim_online(:pilots => true, :atc => true)[:pilots].size.should eq(12)
      "LO".vatsim_online(:atc => false)[:atc].size.should eq(0)
      "LO".vatsim_online(:atc => false)[:pilots].size.should eq(12)
      "WM".vatsim_online(:pilots => false)[:atc].size.should eq(4)
      "LO".vatsim_online(:pilots => false)[:pilots].size.should eq(0)

      "LO".vatsim_online[:pilots].first.callsign.should eq("AMZ1105")
      "WM".vatsim_online[:atc].first.callsign.should eq("WMKK_APP")
      "LO".vatsim_online(:gcmap_width => "400", :gcmap_height => "400")[:pilots].first.gcmap.should eq("http://www.gcmap.com/map?P=LOWW-N45.45676+E12.28972-LIPZ%2C+\"AMZ1105%5Cn1007+ft%5Cn128+kts\"%2B%40N45.45676+E12.28972%0d%0a&MS=wls&MR=120&MX=400x400&PM=b:disc7%2b\"%25U%25+%28N\"")
      "LO".vatsim_online(:gcmap_width => 400, :gcmap_height => 400)[:pilots].first.gcmap.should eq("http://www.gcmap.com/map?P=LOWW-N45.45676+E12.28972-LIPZ%2C+\"AMZ1105%5Cn1007+ft%5Cn128+kts\"%2B%40N45.45676+E12.28972%0d%0a&MS=wls&MR=120&MX=400x400&PM=b:disc7%2b\"%25U%25+%28N\"")
    end

    it "should be case insensitive" do
      gem_data_file
      "wm".vatsim_online[:atc].size.should eq(4)
      "lo".vatsim_online[:pilots].size.should eq(12)
      "wm".vatsim_online(:pilots => true, :atc => true)[:atc].size.should eq(4)
      "lo".vatsim_online(:pilots => true, :atc => true)[:pilots].size.should eq(12)
    end
  end

  describe "vatsim_callsign" do
    it "should work :)" do
      gem_data_file
      "AMZ1105".vatsim_callsign.class.should eq(Array)
      "AMZ1105".vatsim_callsign.size.should eq(1)
      "AMZ1105".vatsim_callsign.first.callsign.should eq("AMZ1105")      
      "BAW".vatsim_callsign.size.should eq(15)
      "BAW".vatsim_callsign.last.callsign.should eq("BAW9DV")      
      "BAW, AMZ1105".vatsim_callsign.size.should eq(16)
      "BAW, QFA".vatsim_callsign.size.should eq(21)
      
    end

    it "should be case insensitive" do
      gem_data_file
      "amz1105".vatsim_callsign.first.callsign.should eq("AMZ1105")
    end
  end

end


describe VatsimTools::Station do

  describe "new object" do
    it "should return proper attributes" do
      gem_data_file
      icao = "WMKK"
      station = VatsimTools::StationParser.new(icao).stations.first
      new_object = VatsimTools::Station.new(station)
      new_object.callsign.should eq("WMKK_APP")
      new_object.name.should eq("Adriel Loke")
      new_object.role.should eq("ATC")
      new_object.frequency.should eq("124.200")
      new_object.rating.should eq("S2")
      new_object.facility.should eq("5")
      new_object.logon.should eq("20120722092836")
      new_object.latitude.should eq("2.93968")
      new_object.latitude_humanized.should eq("N2.93968")
      new_object.longitude.should eq("101.39812")
      new_object.longitude_humanized.should eq("E101.39812")
    end

    it "should parse Ruby time with online_since attr" do
      gem_data_file
      icao = "LBWN"
      station = VatsimTools::StationParser.new(icao).sorted_station_objects[:atc].first
      station.logon.should eq("20120722091954")
      station.online_since.class.should eq(Time)
      station.online_since.utc?.should eq(true)
      station.online_since.should eq("2012-07-22 09:19:54 UTC")
      station.rating.should eq("INS")
    end
  end

  describe "pilot object" do
    it "should contain all attributes" do
      gem_data_file
      icao = "EGLL"
      station = VatsimTools::StationParser.new(icao).stations.first
      new_object = VatsimTools::Station.new(station)
      new_object.callsign.should eq("AAL026")
      new_object.name.should eq("Manuel Santillan MMCU")
      new_object.role.should eq("PILOT")
      new_object.latitude.should eq("44.09780")
      new_object.latitude_humanized.should eq("N44.0978")
      new_object.longitude.should eq("-58.41483")
      new_object.longitude_humanized.should eq("W58.41483")
      new_object.planned_altitude.should eq("370")
      new_object.transponder.should eq("4122")
      new_object.heading.should eq("73")
      new_object.qnh_in.should eq("30.12")
      new_object.qnh_mb.should eq("1019")
      new_object.flight_type.should eq("I")
      new_object.cid.should eq("1210329")
    end

    it "should generate gcmap link" do
      gem_data_file
      icao = "EGLL"
      station = VatsimTools::StationParser.new(icao).stations.first
      new_object = VatsimTools::Station.new(station)
      new_object.gcmap.should eq("http://www.gcmap.com/map?P=KDFW-N44.0978+W58.41483-EGLL%2C+\"AAL026%5Cn37210+ft%5Cn543+kts\"%2B%40N44.0978+W58.41483%0d%0a&MS=wls&MR=120&MX=720x360&PM=b:disc7%2b\"%25U%25+%28N\"")
    end

    it "should handle resized gcmap" do
      gem_data_file
      icao = "EGLL"
      args = {}
      args[:gcmap_width] = "400"
      args[:gcmap_height] = "400"
      station = VatsimTools::StationParser.new(icao).stations.first
      new_object = VatsimTools::Station.new(station, args)
      new_object.gcmap_width.should eq(400)
      new_object.gcmap_height.should eq(400)
      new_object.gcmap.should eq("http://www.gcmap.com/map?P=KDFW-N44.0978+W58.41483-EGLL%2C+\"AAL026%5Cn37210+ft%5Cn543+kts\"%2B%40N44.0978+W58.41483%0d%0a&MS=wls&MR=120&MX=400x400&PM=b:disc7%2b\"%25U%25+%28N\"")

    end

  end

  describe "atc object" do
    it "should handle regular and humanized atis" do
      gem_data_file
      icao = "LBWN"
      station = VatsimTools::StationParser.new(icao).sorted_station_objects[:atc].first
      station.logon.should eq("20120722091954")
      station.rating.should eq("INS")
      station.atis.should eq("$ voice2.vacc-sag.org/lfmn_app. Nice Approach. Charts at www.tinyurl.com/chartsfr. Visit www.vatfrance.org")
      station.atis_message.should eq("Nice Approach<br />Charts at www.tinyurl.com/chartsfr<br />Visit www.vatfrance.org")
    end

     it "should handle no ATC remark" do
      gem_data_file
      icao = "NZAA"
      station = VatsimTools::StationParser.new(icao).sorted_station_objects[:atc].first
      station.atis.should eq("$ rw1.vatpac.org/nzaa_twr")
      station.atis_message.should eq("No published remark")
    end
  end

end


lib/vatsim_online/station_parser.rb
module VatsimTools

  class StationParser

    %w{tmpdir csv}.each { |lib| require lib }
    require_relative "data_downloader"
    require_relative "station"

    attributes = %w{role icao excluded gcmap_width gcmap_height}
    attributes.each {|attribute| attr_accessor attribute.to_sym }

    LOCAL_DATA = "#{Dir.tmpdir}/vatsim_data.txt"

    def initialize(icao, args = nil)
      VatsimTools::DataDownloader.new
      args.class == Hash ? @role = determine_role(args) : @role = "all"
      if icao == "ALL"
        @icao = nil
      else
        @icao = icao.upcase.split(',').each {|s| s.strip!}
      end
      @excluded = args[:exclude].upcase if args && args[:exclude]
      @gcmap_width = args[:gcmap_width] if args && args[:gcmap_width]
      @gcmap_height = args[:gcmap_height] if args && args[:gcmap_height]
    end

    def determine_role(args)
      args[:atc] == false ? role = "pilot" : role = "all"
      args[:pilots] == false ? role = "atc" : role = role
      role = "all" if args[:pilots] == false && args[:atc] == false
      role
    end

    def stations
      stations = []
      CSV.foreach(LOCAL_DATA, :col_sep =>':') do |row|
        callsign, origin, destination, client = row[0].to_s, row[11].to_s, row[13].to_s, row[3].to_s
        unless @icao
          stations << row if (client == "ATC") unless @role == "pilot"
          stations << row if (client == "PILOT") unless @role == "atc"
        else
          for icao in @icao
            stations << row if (callsign[0...icao.length] == icao && client == "ATC") unless @role == "pilot"
            stations << row if (origin[0...icao.length] == icao || destination[0...icao.length] == icao) unless @role == "atc"
          end
        end
      end
      stations
    end

    def station_objects
      station_objects= []
      args = {}
      args[:gcmap_width] = @gcmap_width if @gcmap_width
      args[:gcmap_height] = @gcmap_height if @gcmap_height
      stations.each {|station| station_objects << VatsimTools::Station.new(station, args) }
      station_objects
    end

    def sorted_station_objects
      atc = []; pilots = []; arrivals = []; departures = []
      station_objects.each {|sobj| sobj.role == "ATC" ? atc << sobj : pilots << sobj}
      if @icao
        for icao in @icao
          for pilot in pilots
            departures << pilot if pilot.origin[0...icao.length] == icao
            arrivals << pilot if pilot.destination[0...icao.length] == icao
          end
        end
      end
      atc.delete_if {|a| @excluded && a.callsign[0...@excluded.length] == @excluded }
      {:atc => atc, :pilots => pilots, :arrivals => arrivals, :departures => departures}
    end

  end

end


lib/vatsim_online/callsign_parser.rb
module VatsimTools

  class CallsignParser

    %w{tmpdir csv}.each { |lib| require lib }
    require_relative "data_downloader"
    require_relative "station"

    attributes = %w{role callsign gcmap_width gcmap_height}
    attributes.each {|attribute| attr_accessor attribute.to_sym }

    LOCAL_DATA = "#{Dir.tmpdir}/vatsim_data.txt"

    def initialize(callsign, args = nil)
      VatsimTools::DataDownloader.new
      # args.class == Hash ? @role = determine_role(args) : @role = "all"
      @callsign = callsign.upcase.split(',').each {|s| s.strip!}
      # @excluded = args[:exclude].upcase if args && args[:exclude]
      @gcmap_width = args[:gcmap_width] if args && args[:gcmap_width]
      @gcmap_height = args[:gcmap_height] if args && args[:gcmap_height]
    end

    # def determine_role(args)
    #   args[:atc] == false ? role = "pilot" : role = "all"
    #   args[:pilots] == false ? role = "atc" : role = role
    #   role = "all" if args[:pilots] == false && args[:atc] == false
    #   role
    # end


    def stations
      stations = []
      CSV.foreach(LOCAL_DATA, :col_sep =>':') do |row|
        callsign, origin, destination, client = row[0].to_s, row[11].to_s, row[13].to_s, row[3].to_s
        for cs in @callsign
          stations << row if callsign[0...cs.length] == cs # && client == "ATC") unless @role == "pilot"
          # stations << row if (origin[0...icao.length] == icao || destination[0...icao.length] == icao) unless @role == "atc"
        end
      end
      stations
    end

    def station_objects
      station_objects= []
      args = {}
      args[:gcmap_width] = @gcmap_width if @gcmap_width
      args[:gcmap_height] = @gcmap_height if @gcmap_height
      stations.each {|station| station_objects << VatsimTools::Station.new(station, args) }
      station_objects
    end

  end

end


lib/vatsim_online/station.rb
# encoding: utf-8
module VatsimTools
  class Station
    require "gcmapper"

    attributes = %w{callsign name role frequency altitude groundspeed aircraft
      origin destination rating facility remarks route atis logon latitude longitude
      planned_altitude transponder heading qnh_in qnh_mb flight_type cid gcmap
      latitude_humanized longitude_humanized online_since gcmap_width gcmap_height
      atis_message}
    attributes.each {|attribute| attr_accessor attribute.to_sym }

    def initialize(station, args = nil)

      @callsign, @cid,  @name, @role, @frequency, @latitude, @longitude,  @altitude, @groundspeed, @aircraft, @origin,
        @planned_altitude, @destination, @transponder, @facility, @flight_type, @remarks, @route, @logon, @heading,
        @qnh_in, @qnh_mb = station[0], station[1], station[2], station[3], station[4], station[5], station[6], station[7],
        station[8], station[9], station[11], station[12], station[13], station[17], station[18], station[21], station[29],
        station[30], station[37], station[38], station[39], station[40]

      @atis = atis_cleaner(station[35]) if station[35]
      @rating = humanized_rating(station[16])
      @latitude_humanized = latitude_parser(station[5].to_f)
      @longitude_humanized = longitude_parser(station[6].to_f)
      @online_since = utc_logon_time if @logon
      @gcmap_width = args[:gcmap_width].to_i if args && args[:gcmap_width]
      @gcmap_height = args[:gcmap_height].to_i if args && args[:gcmap_height]
      @gcmap = gcmap_generator
      @atis_message = construct_atis_message(station[35]) if station[35]
    end

  private

    def gcmap_generator
      return "No map for ATC stations" if @role != "PILOT"
      construct_gcmap_url.gcmap(:width => @gcmap_width, :height => @gcmap_height)
    end

    def construct_gcmap_url
      if @origin && @latitude_humanized && @longitude_humanized && @destination
        route = @origin.to_s + "-" + @latitude_humanized.to_s + "+" + @longitude_humanized.to_s + "-" + @destination.to_s
        route += "%2C+\"" + @callsign.to_s + "%5Cn" + @altitude.to_s + "+ft%5Cn" + @groundspeed.to_s + "+kts"
        route += "\"%2B%40" + @latitude_humanized.to_s + "+" + @longitude_humanized.to_s
      else
        route = "Position undetermined"
      end
      route
    end

    def latitude_parser(lat)
      lat > 0 ? hemisphere = "N" : hemisphere = "S"
      hemisphere + lat.abs.to_s
    end

    def longitude_parser(lon)
      lon > 0 ? hemisphere = "E" : hemisphere = "W"
      hemisphere + lon.abs.to_s
    end

    def atis_cleaner(raw_atis)
      raw_atis.gsub(/[\^]/, '. ')
    end

    def utc_logon_time
      Time.parse ("#{@logon[0...4]}-#{@logon[4...6]}-#{@logon[6...8]} #{@logon[8...10]}:#{@logon[10...12]}:#{@logon[12...14]} UTC")
    end

    def humanized_rating(rating_number)
      case rating_number
        when "0" then "Suspended"
        when "1" then "OBS"
        when "2" then "S1"
        when "3" then "S2"
        when "4" then "S3"
        when "5" then "C1"
        when "7" then "C3"
        when "8" then "INS"
        when "10" then "INS+"
        when "11" then "Supervisor"
        when "12" then "Administrator"
        else
          "UNK"
      end
    end

    def construct_atis_message(raw_atis)
      message = raw_atis.gsub(/[\^]/, '<br />')
      message.index('>') ? message = message[message.index('>')+1...message.length] : message = "No published remark"
    end

  end
end
